File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution{
public:
    long long check(vector<int> &nums, int h, int m){
        long long ans = 0;

        for(int num : nums){
            ans += num / m;

            if(num % m != 0) ans++;
            if(ans > h) break;
        }

        return ans;
    }
    int minEatingSpeed(vector<int> nums, int h){
        // Code Here
        ios_base::sync_with_stdio(false);
        int ans = INT_MAX;

        int i = 1, j = *max_element(nums.begin(), nums.end());

        while(i <= j){
            int m = i + (j - i) / 2;

            long long x = check(nums, h, m);

            if(x <= h){
                ans = min(ans, m);
                j = m - 1;
            }
            else if(x > h) i = m + 1;
        }
        
        return ans;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int check(int mid, int[] piles, int h){
        int ans = 0;

        for(int i = 0; i < piles.length; i++){
            ans += piles[i] / mid;
            if(piles[i] % mid != 0){
                ans += 1;
            }

            if(ans > h) return 0;
        }

        return 1;


    }
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int maxValue = piles[0];

        for(int i = 0;  i< piles.length; i++){
            if(piles[i] > maxValue){
                maxValue = piles[i];
            }
        }

        int right = maxValue;

        int ans = -1;


        while(left <= right){
            int mid = left + (right - left) / 2;

            int flg = check(mid, piles, h);

            if(flg == 0){
               left = mid + 1;
            }
            else{
                ans = mid;
                right = mid - 1;
            }


        }


        return ans;
    }
}